File: src/experimental/oleksandr/triple.cpp
/*
  Name: triple.h
  Copyright: 2006 Nanorex, Inc.  All rights reserved.
  Author: Oleksandr Shevchenko
  Description: coordinate 3D point management class 
*/

#include "triple.h"

//----------------------------------------------------------------------------
// Closest()
//
// closest point on edge
//
double Triple::Closest(
	const Triple & vertex0,
	const Triple & vertex1,
	Triple & p) const
{
	Triple edge(vertex0,vertex1);
	double t = ((*this-vertex0)%edge)/edge.Len2();
	if (t < 0.0)
	{
		p = vertex0;
		t = 0.0;
	}
	else if (t > 1.0)
	{
		p = vertex1;
		t = 1.0;
	}
	else
	{
		p = vertex0 + t*edge;
	}
	return t;
}

//----------------------------------------------------------------------------
// Greatest()
//
// calculate greatest value
//
int Triple::Greatest(
	double & value) const
{
	if (X() > Y())
	{
		if(X() > Z())
		{
	//    x is greatest
			value = X();
			return (0);
		}
		else
		{
	//    z is greatest
			value = Z();
			return (2);
		}
	}
	else
	{
		if(Y() > Z())
		{
	//    y is greatest
			value = Y();
			return (1);
		}
		else
		{
	//    z is greatest
			value = Z();
			return (2);
		}
	}
	return (2);
}

//----------------------------------------------------------------------------
// Len()
//
//  vector length
//
double Triple::Len() const
{
	return sqrt(X()*X() + Y()*Y() + Z()*Z());
}

//----------------------------------------------------------------------------
// Len2()
//
//  square of vector length
//
double Triple::Len2() const
{
	return (X()*X() + Y()*Y() + Z()*Z());
}

//----------------------------------------------------------------------------
// Normalize()
//
//  normalize vector to unit length
//
Triple & Triple::Normalize()
{
	double length= Len();
	*this /= length;
	return *this;
}

//----------------------------------------------------------------------------
// Normalize()
//
//  normalizes vector to unit length
//
Triple & Triple::Normalize(
	double &length)
{
	length = Len();
	*this /= length;
	return *this;
}

//----------------------------------------------------------------------------
// Orthogonal()
//
// get orthogonal axes
//
void Triple::Orthogonal(
	Triple & ox,
	Triple & oy) const
{
	double ix = 0.0, iy = 0.0;
	
	if( fabs(X())>fabs(Z()) && fabs(X()) > fabs(Y()) )
		iy = 1.0;
	else
		ix = 1.0;
	
	oy = Triple(iy*Z(), -ix*Z(), ix*Y()-iy*X());
	oy.Normalize();
	
	ox = oy * (*this);
	ox.Normalize();
}

File: cad/src/experimental/oleksandr/Triple.py
# Copyright 2006-2007 Nanorex, Inc.  See LICENSE file for details.

__author__ = 'oleksandr'

import types, operator, math

"""3D vector representation"""

class Triple:

    def __init__(self, *args):
        """vector constructor"""
        self.x = 0
        self.y = 0
        self.z = 0
        if len(args) == 0:
            pass
        if len(args) == 1:
            if isinstance(args[0], types.InstanceType):
                self.x = args[0].x
                self.y = args[0].y
                self.z = args[0].z
            elif isinstance(args[0], types.ListType):
                self.x = args[0][0]
                self.y = args[0][1]
                self.z = args[0][2]
            else:
                self.x = args[0]
                self.y = args[0]
                self.z = args[0]
        if len(args) == 2:
            self.x = args[1][0] - args[0][0]
            self.y = args[1][1] - args[0][1]
            self.z = args[1][2] - args[0][2]
        if len(args) == 3:
            self.x, self.y, self.z = args

    def __str__(self):
        """returns the triple in a textual form"""
        s = ""
        s += "%s " % self.x
        s += "%s " % self.y
        s += "%s " % self.z
        return s

    def Len2(self):
        """square of vector length"""
        return self.x * self.x + self.y * self.y + self.z * self.z

    def Len(self):
        """vector length"""
        return math.sqrt(self.Len2())

    def Normalize(self):
        """normalizes vector to unit length"""
        length = self.Len();
        self.x /= length
        self.y /= length
        self.z /= length
        return self

    def Greatest(self):
        """calculate greatest value"""
        if self.x > self.y:
            if self.x > self.z:
                return self.x
            else:
                return self.z
        else:
            if self.y > self.z:
                return self.y
            else:
                return self.z

    def __add__( self, rhs ):
        """operator a + b"""
        t = Triple(rhs)
        return Triple(self.x + t.x, self.y + t.y, self.z + t.z)

    def __radd__( self, lhs ):
        """operator b + a"""
        t = Triple(lhs)
        t.x += self.x
        t.y += self.y
        t.z += self.z
        return t

    def __sub__( self, rhs ):
        """operator a - b"""
        t = Triple(rhs)
        return Triple(self.x - t.x, self.y - t.y, self.z - t.z)

    def __rsub__( self, lhs ):
        """operator b - a"""
        t = Triple(lhs)
        t.x -= self.x
        t.y -= self.y
        t.z -= self.z
        return t

    def __mul__( self, rhs ):
        """operator a * b"""
        t = Triple(rhs)
        return Triple(self.x * t.x, self.y * t.y, self.z * t.z)

    def __rmul__( self, lhs ):
        """operator b * a"""
        t = Triple(lhs)
        t.x *= self.x
        t.y *= self.y
        t.z *= self.z
        return t

    def __div__( self, rhs ):
        """operator a / b"""
        t = Triple(rhs)
        return Triple(self.x / t.x, self.y / t.y, self.z / t.z)

    def __rdiv__( self, lhs ):
        """operator b / a"""
        t = Triple(lhs)
        t.x /= self.x
        t.y /= self.y
        t.z /= self.z
        return t

    def __mod__( self, rhs ):
        """operator a % b (scalar product)"""
        r = self.x * rhs.x + self.y * rhs.y + self.z * rhs.z
        return r

    def __neg__( self):
        """operator -a"""
        return Triple(-self.x, -self.y, -self.z)

def Dot(lhs, rhs ):
    """function dot(a, b) (scalar product)"""
    r = lhs.x * rhs.x + lhs.y * rhs.y + lhs.z * rhs.z
    return r

